torch/csrc/jit/codegen/cuda/fusion.cpp

```diff
@@ -57,6 +59,103 @@ std::unordered_set<Val*> InputsOf::output(Fusion* fusion, Val* output_) {
   return io.inputs;
 }
 
+void swap(Fusion& a, Fusion& b) noexcept {
+  using std::swap;
+
+  // Swap the content
+  swap(a.val_set_, b.val_set_);
+  swap(a.expr_set_, b.expr_set_);
+  swap(a.val_deque_, b.val_deque_);
+
+  swap(a.val_type_name_map_, b.val_type_name_map_);
+  swap(a.val_name_counter_, b.val_name_counter_);
+  swap(a.expr_name_counter_, b.expr_name_counter_);
+
+  swap(a.origin_, b.origin_);
+  swap(a.uses_, b.uses_);
+  swap(a.values_map_, b.values_map_);
+
+  swap(a.inputs_, b.inputs_);
+  swap(a.outputs_, b.outputs_);
+
+  // Fixup the Statement::fusion_ links for a
+  for (auto val : a.val_set_) {
+    val->fusion_ = &a;
+  }
+  for (auto expr : a.expr_set_) {
+    expr->fusion_ = &a;
+  }
+
+  // Fixup the Statement::fusion_ links for b
+  for (auto val : b.val_set_) {
+    val->fusion_ = &b;
+  }
+  for (auto expr : b.expr_set_) {
+    expr->fusion_ = &b;
+  }
+}
+
+Fusion::Fusion(const Fusion& other) {
+  IrCloner ir_cloner(this);
+
+  for (auto val : other.val_set_) {
+    val_set_.insert(ir_cloner.clone(val));
+  }
+
+  for (auto expr : other.expr_set_) {
+    expr_set_.insert(ir_cloner.clone(expr));
+  }
+
+  for (auto val : other.val_deque_) {
+    val_deque_.push_back(ir_cloner.clone(val));
+  }
+
+  val_type_name_map_ = other.val_type_name_map_;
+  val_name_counter_ = other.val_name_counter_;
+  expr_name_counter_ = other.expr_name_counter_;
+
+  for (const auto& kv : other.origin_) {
+    auto val = ir_cloner.clone(kv.first);
+    auto expr = ir_cloner.clone(kv.second);
+    origin_.insert({val, expr});
+  }
+
+  for (const auto& kv : other.uses_) {
+    auto val = ir_cloner.clone(kv.first);
+    std::unordered_set<Expr*> val_uses;
+    for (auto expr : kv.second) {
+      val_uses.insert(ir_cloner.clone(expr));
+    }
+    uses_.insert({val, std::move(val_uses)});
+  }
+
+  for (const auto& kv : other.values_map_) {
+    auto from_val = ir_cloner.clone(kv.first);
+    auto to_val = ir_cloner.clone(kv.second);
+    values_map_.insert({from_val, to_val});
+  }
+
+  inputs_ = ir_cloner.clone(other.inputs_);
+  outputs_ = ir_cloner.clone(other.outputs_);
+}
+
+Fusion::Fusion(Fusion&& other) noexcept {
```

Comment:
Any specific use for move constructors you have in mind, or just wanted to support them?

Reply:
For containers where copies are expensive, move operations can make a significant improvement in the quality of life. You can easily say

`Fusion fusion = createSchedule(base_fusion);`

Instead of being forced to resort to ugly contortions (ex. out parameters).

In general, and I found this to be an important guideline, implementing resource ownership is much easier as an all-or-nothing proposition. This rule even has it's own name: https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)